walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,67 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.List;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import com.google.common.collect.ImmutableList; 
 
+/** Finds list of all potential calendar time slots to schedule a meeting with the given 
+ * MeetingRequest constraints. Accounts for optional attendees when possible. */
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ImmutableList<TimeRange> conflicting = getRelevantEvents(request.getAttendees(), events);
+    ImmutableList<TimeRange> times = getMeetingTimes(request.getDuration(), conflicting);
+    
+    Iterable<String> allAttendeesIter = 
+        Iterables.unmodifiableIterable(Iterables.concat(request.getAttendees(),
+                                                        request.getOptionalAttendees()));
+    Collection<String> allAttendees = Lists.newArrayList(allAttendeesIter);
+    
+    ImmutableList<TimeRange> optionalConflicting = getRelevantEvents(allAttendees, events);
+    ImmutableList<TimeRange> timesWithOptional = getMeetingTimes(request.getDuration(), optionalConflicting);
+
+    if (timesWithOptional.size() == 0 && request.getAttendees().size() != 0) {
+      return times;
+    } else {
+      return timesWithOptional;
+    }
+  }
+
+  // Returns list of events that at least one of given attendees is going to.
+  private ImmutableList<TimeRange> getRelevantEvents(Collection<String> attendees, Collection<Event> events) {
```

Comment:
nit: let's rename this method since it doesn't return Events, but rather TimeRange. Maybe something like getAttendeeTimeRanges?